eclipse/src/saros/ui/model/roster/RosterContentProvider.java

```diff
@@ -1,216 +1,71 @@
-        }
-
-        @Override
-        public void entriesAdded(Collection<String> addresses) {
-          ViewerUtils.refresh(viewer, true);
-          querySarosSupport(addresses);
-        }
-      };
+  private Viewer viewer;
+  private final IContactsUpdate contactsUpdate =
+      (contact, type) -> ViewerUtils.refresh(viewer, true);
 
   public RosterContentProvider() {
     SarosPluginContext.initComponent(this);
-
-    discoveryManager.addDiscoveryManagerListener(discoveryManagerListener);
+    contactsService.addListener(contactsUpdate);
   }
 
   @Override
   public void inputChanged(Viewer viewer, Object oldInput, Object newInput) {
     this.viewer = viewer;
-
-    if (oldInput instanceof Roster) ((Roster) oldInput).removeRosterListener(rosterListener);
-
-    roster = null;
-
-    if (newInput instanceof Roster) {
-      roster = (Roster) newInput;
-      roster.addRosterListener(rosterListener);
-
-      for (final RosterEntry entry : roster.getEntries())
-        querySarosSupport(Collections.singletonList(entry.getUser()));
-    }
   }
 
   @Override
   public void dispose() {
-    if (roster != null) roster.removeRosterListener(rosterListener);
-
-    discoveryManager.removeDiscoveryManagerListener(discoveryManagerListener);
-
-    roster = null;
-    discoveryManager = null;
+    contactsService.removeListener(contactsUpdate);
   }
 
   /**
-   * Returns {@link RosterGroup}s followed by {@link RosterEntry}s which don't belong to any {@link
-   * RosterGroup}.
+   * Returns ContactGroups followed by Contacts which don't belong to any Group.
+   *
+   * @param unused is unused
    */
   @Override
-  public Object[] getElements(Object inputElement) {
-
-    if (!(inputElement instanceof Roster)) return new Object[0];
-
-    Roster inputRoster = (Roster) inputElement;
-    final List<Object> elements = new ArrayList<Object>();
-
-    /*
-     * always show contacts that support Saros regardless of any other
-     * source, i.e the user is online with:
-     *
-     * alice@foo/Saros (<- has Saros support)
-     *
-     * and
-     *
-     * alice@foo/Pidgen
-     *
-     * so always display alice@foo/Saros
-     */
-
-    for (RosterGroup group : inputRoster.getGroups()) {
-      elements.add(
-          new RosterGroupElement(
-              group,
-              filterRosterEntryElements(createRosterEntryElements(group.getEntries()))
-                  .toArray(new RosterEntryElement[0])));
-    }
-
-    elements.addAll(
-        filterRosterEntryElements(createRosterEntryElements(inputRoster.getUnfiledEntries())));
-
-    return elements.toArray();
-  }
-
-  private List<RosterEntryElement> createRosterEntryElements(
-      final Collection<RosterEntry> entries) {
-
-    final List<RosterEntryElement> elements = new ArrayList<RosterEntryElement>();
-
-    for (final RosterEntry entry : entries)
-      elements.add(createRosterEntryElement(new JID(entry.getUser())));
+  public Object[] getElements(Object unused) {
```

Comment:
For clarification, a content provider is responsible for generating new data. However it is not responsible for the source of the data. Just because the ContactsService is a single is no excuse. In general you can expect that this provider can handle different IContactsService in the future. The other option would to write one for each IContactService implementation. Last option would require a bunch of if/elseif/else stuff in every widget that uses a viewer with this content provider.

Reply:
Thanks! Yes, I wanted to avoid more changes, but ok... :D
Now changed only a minimum and I don't like this UI code with its castings, but it works.
More UI rework should be part of another pr.